lib/aws/client.ex

```diff
@@ -194,12 +194,97 @@ defmodule AWS.Client do
     %{client | http_client: http_client}
   end
 
+  @doc """
+  Makes a HTTP request using the specified client.
+
+  `opts` may contain the keyword `:enable_retries?` that enables request retries on known
+  errors such as 500s.
+  The keyword `:retry_opts` can further supply the arguments: `:max_retries`, `:base_sleep_time` &
+  `:cap_sleep_time` that control the retry mechanism. This uses exponential backoff with jitter to
+  sleep in between retry attempts. Default values are:
+  `retry_opts: [max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000]`
+
+  See "FullJitter" at: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+  ### Examples
+
+      iex> AWS.Client.request(client, :post, url, payload, headers, options)
+      {:ok, %{status_code: 200, body: body}}
+
+      iex> AWS.Client.request(client, :post, url, payload, headers, enable_retries?: true)
+      {:ok, %{status_code: 200, body: body}}
+  """
   def request(client, method, url, body, headers, opts \\ []) do
+    # Pop off all retry-related options from opts, so they aren't passed to the HTTP client.
+    {enable_retries?, opts} = Keyword.pop(opts, :enable_retries?, false)
```

Comment:
I know I'm late in the review, but what about removing this `:enable_retries?` option and use the `max_retries` as a way to enable or disable this? We could set `max_retries` to `0` in order to disable. WDYT?

If this is too much for now, we can refactor in another PR.

Reply:
I agree the API would be nicer, but I was afraid to shadow options for the clients. `max_retries` seems like such a common option, that's why I wrapped it in `retry_opts`.
AFAIK neither Finch or Hackney uses `max_retries`, but Req does: https://hexdocs.pm/req/0.2.1/Req.html#retry/2
Not sure about other clients.
I also considering pushing the responsibility down into the AWS.HTTPClients instead. That would also remove the hackney specific clauses here: https://github.com/aws-beam/aws-elixir/pull/182/files/f4783edac5f93637cff434e40a1532f2e4d9d925#diff-5a31358e99eb904d8d9019d153ff5dec3f14890d446a070ec1010934c71fb910R280

What do you think?

Reply:
OK, I don't have a strong opinion about this. So I'm fine to keep as it is :)

Reply:
Keeping as is, to get it shipped then